elastic/elasticsearch
# EIS integration

This PR integrates EIS (Elastic Inference Service) with Elasticsearch behind a feature flag.

### Useful ES commands:

- Running Elasticsearch (inside of the ES root directory):
  - Run ES: `./gradlew run`
  - Run ES in debug mode: `./gradlew run --debug-jvm` (you can attach a debugger on the debug port, which will be logged)
  - Run ES and set EIS gateway URL via CLI: `./gradlew run -Dtests.es.xpack.inference.eis.gateway.url=http://localhost:8000`
- Running tests:
  - Run tests inside inference API plugin: `./gradlew ':x-pack:plugin:inference:test'`
  - Run a specific test class inside inference API plugin: `./gradlew ':x-pack:plugin:inference:test' --tests "org.elasticsearch.xpack.inference.external.response.elastic.ElasticInferenceServiceSparseEmbeddingsResponseEntityTests"` (specify `--tests` and package + class name)
  - Run a specific test method inside Inference API plugin: `./gradlew ':x-pack:plugin:inference:test' --tests "org.elasticsearch.xpack.inference.external.response.elastic.ElasticInferenceServiceSparseEmbeddingsResponseEntityTests.testSparseEmbeddingsResponse_SingleEmbeddingInData_NoMeta_NoTruncation"` (specify `--tests` and package + class + method name)
- Check style/formatting:
  - Check style in inference API plugin: `./gradlew ':x-pack:plugin:inference:checkstyleTest'`
  - Apply spotless/formatting: `./gradlew spotlessApply` 

### Testing locally:
- Start `eis-model-server` (or `eis-gateway`, if `ELSERv2` endpoint is integrated) on port `{PORT}`
- Start ES with a configured `eis-gateway`: `./gradlew run -Dtests.es.xpack.inference.eis.gateway.url=http://localhost:{PORT}`
- Create an EIS inference endpoint:
```
PUT {ES_HOST}/_inference/sparse_embedding/eis

{
    "service": "elastic",
    "service_settings": {
        "model_id": ".elser_model_2"
    }
}
```
- Perform inference (single embedding in a list):
```
POST {ES_HOST}/_inference/sparse_embedding/eis

{
    "input": "A blue sky"
}
```
- Perform inference (multiple embeddings):
```
POST {ES_HOST}/_inference/sparse_embedding/eis

{
    "input": [
        "Embed this text",
        "Embed this text, too",
        "This text should also be embedded"
    ]
}
```

### Testing in serverless:
- Enable the feature flag for your project or a whole environment via a PR in our corresponding gitops repo via `jvmOptions`: `"-Des.eis_feature_flag_enabled=true"`
- Repeat steps from `Testing locally`

### TODOs:
- [x] Write tests for `ElasticInferenceService`
- [x] Write tests for `ElasticInferenceServiceActionCreator`
- [x] Write tests for `ElasticInferenceServiceResponseHandler`
- [x] Write tests for `ElasticInferenceServiceSparseEmbeddingsRequest`
- [x] Implement `checkModelConfig` in `ElasticInferenceService`
- [x] Implement `doChunkedInfer` in `ElasticInferenceService`
- [x] Implement truncation in `ElasticInferenceServiceSparseEmbeddingsRequest`
- [x] Add docs for `ElasticInferenceServiceFeature`
- [x] Handle error codes specified in inference Task API spec inside `ElasticInferenceServiceResponseHandler`
- (There might be some additional smaller `TODOs`, which I forgot here, I usually mark them with `//TODO:`)

### When ready for review:
- Mark ready for review
- Add labels (this will ping the ML team for reviews):
  - `>non-issue`
  -  `:ml`
  - `Team:ML`

### Out of scope for this PR:
- "Always-on" experience (performing inference without creating an endpoint for service `elastic` upfront)
- Rate limiting
- Auth/Secret Settings
